tools/circt-verilog/circt-verilog.cpp

```diff
@@ -367,24 +406,73 @@ static LogicalResult executeWithSources(MLIRContext *context,
   }
 
   // Print the final MLIR.
+  auto outputTimer = ts.nest("MLIR Printer");
   module->print(outputFile->os());
   outputFile->keep();
   return success();
 }
 
 static LogicalResult execute(MLIRContext *context) {
+  // Abort if there are no input files to be processed.
+  if (opts.inputFilenames.empty()) {
+    WithColor::error() << "no input files\n";
+    return failure();
+  }
+
+  // Auto-detect the input format if it was not explicitly specified.
+  if (opts.format.getNumOccurrences() == 0) {
+    std::optional<Format> detectedFormat = std::nullopt;
+    for (const auto &inputFilename : opts.inputFilenames) {
+      std::optional<Format> format = std::nullopt;
+      auto name = StringRef(inputFilename);
+      if (name.ends_with(".v") || name.ends_with(".sv") ||
+          name.ends_with(".vh") || name.ends_with(".svh"))
+        format = Format::SV;
+      else if (name.ends_with(".mlir") || name.ends_with(".mlirbc"))
+        format = Format::MLIR;
+      if (!format)
+        continue;
+      if (detectedFormat && format != detectedFormat) {
+        detectedFormat = std::nullopt;
+        break;
+      }
+      detectedFormat = format;
+    }
+    if (!detectedFormat) {
+      WithColor::error() << "cannot auto-detect input format; use --format\n";
+      return failure();
+    }
+    opts.format = *detectedFormat;
+  }
+
   // Open the input files.
   llvm::SourceMgr sourceMgr;
   for (const auto &inputFilename : opts.inputFilenames) {
     std::string errorMessage;
     auto buffer = openInputFile(inputFilename, &errorMessage);
     if (!buffer) {
-      llvm::errs() << errorMessage << "\n";
+      WithColor::error() << errorMessage << "\n";
       return failure();
     }
     sourceMgr.AddNewSourceBuffer(std::move(buffer), llvm::SMLoc());
   }
 
+  // Register the dialects.
+  // clang-format off
+  context->loadDialect<
+    comb::CombDialect,
+    debug::DebugDialect,
+    func::FuncDialect,
+    hw::HWDialect,
+    llhd::LLHDDialect,
+    moore::MooreDialect,
+    scf::SCFDialect,
+    seq::SeqDialect,
+    sim::SimDialect,
+    verif::VerifDialect
+  >();
+  // clang-format on
```

Comment:
Nit: I'm not really a fan of switching off clang-format. If you want to make it easier to understand/read then maybe group them with line comments by abstraction level or sth?

Reply:
Hmmm, the way clang format bunches these up is kind of annoying, because you can't clearly see if something gets added to the list. It's always just a blob of delta. :thinking: